provider/openstack/local_test.go

```diff
@@ -2548,6 +2554,131 @@ func (s *localServerSuite) TestUpdateGroupController(c *gc.C) {
+		res, ok := args[0].(*identityservice.AccessResponse)
+		c.Assert(ok, jc.IsTrue)
+		var filtered []identityservice.V2Service
+		for _, service := range res.Access.ServiceCatalog {
+			if service.Type != "network" {
+				filtered = append(filtered, service)
+			}
+		}
+		res.Access.ServiceCatalog = filtered
+		return nil
+	})
+	s.AddCleanup(func(c *gc.C) { cleanup() })
+
+	cl := client.NewClient(s.cred, identity.AuthUserPass, nil)
+	err := cl.Authenticate()
+	c.Assert(err, jc.ErrorIsNil)
+	containerURL, err := cl.MakeServiceURL("object-store", "", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	attrs := coretesting.FakeConfig().Merge(coretesting.Attrs{
+		"name":               "sample-no-neutron",
+		"type":               "openstack",
+		"auth-mode":          "userpass",
+		"agent-version":      coretesting.FakeVersionNumber.String(),
+		"agent-metadata-url": containerURL + "/juju-dist-test/tools",
+		"image-metadata-url": containerURL + "/juju-dist-test",
+		"authorized-keys":    "fakekey",
+	})
+	s.PatchValue(&jujuversion.Current, coretesting.FakeVersionNumber)
+	// For testing, we create a storage instance to which is uploaded tools and image metadata.
+	env, err := bootstrap.Prepare(
+		envtesting.BootstrapContext(c),
+		jujuclient.NewMemStore(),
+		prepareParams(attrs, s.cred),
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	s.env = env
+	s.toolsMetadataStorage = openstack.MetadataStorage(s.env)
+	// Put some fake metadata in place so that tests that are simply
+	// starting instances without any need to check if those instances
+	// are running can find the metadata.
+	envtesting.UploadFakeTools(c, s.toolsMetadataStorage, s.env.Config().AgentStream(), s.env.Config().AgentStream())
+	s.imageMetadataStorage = openstack.ImageMetadataStorage(s.env)
+	openstack.UseTestImageData(s.imageMetadataStorage, s.cred)
+	s.storageAdapter = makeMockAdapter()
+	overrideCinderProvider(c, &s.CleanupSuite, s.storageAdapter)
+}
+
+func (s *noNeutronSuite) TearDownTest(c *gc.C) {
+	if s.imageMetadataStorage != nil {
+		openstack.RemoveTestImageData(s.imageMetadataStorage)
+	}
+	if s.toolsMetadataStorage != nil {
+		envtesting.RemoveFakeToolsMetadata(c, s.toolsMetadataStorage)
+	}
+	s.srv.stop()
+	s.BaseSuite.TearDownTest(c)
+}
+
+func (s *noNeutronSuite) TestUpdateGroupControllerNoNeutron(c *gc.C) {
+	client := openstack.GetNovaClient(s.env)
+	_, err := client.CreateSecurityGroup("unrelated", "you know, for tests!")
+	c.Assert(err, jc.ErrorIsNil)
+	err = bootstrapEnv(c, s.env)
+	c.Assert(err, jc.ErrorIsNil)
+
+	groupNamesBefore := set.NewStrings(getNovaSecurityGroupNames(c, client)...)
+	c.Assert(groupNamesBefore, gc.DeepEquals, set.NewStrings(
+		"default",
+		"unrelated",
+		"juju-deadbeef-1bad-500d-9000-4b1d0d06f00d-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+		"juju-deadbeef-1bad-500d-9000-4b1d0d06f00d-deadbeef-0bad-400d-8000-4b1d0d06f00d-0",
+	))
+
+	firewaller := openstack.GetFirewaller(s.env)
+	err = firewaller.UpdateGroupController("aabbccdd-eeee-ffff-0000-0123456789ab")
```

Comment:
One thing that this clearly indicates to me, is that we messed up our layering, but it isn't your fault.
goose should *never* have a constant with 'juju' in the name.

Reply:
Oh, it's not called out (I'll add a comment) but those `juju-` groups are added by the `bootstrapEnv` call - nothing in goose is creating them.